praw/models/reddit/rules.py

```diff
@@ -0,0 +1,419 @@
+        for rule in self.subreddit.rules:
+            if rule.short_name == self.short_name:
+                self.__dict__.update(rule.__dict__)
+                self._fetched = True
+                return
+        raise ClientException(
+            "r/{} does not have the rule {}".format(
+                self.subreddit, self.short_name
+            )
+        )
+
+
+class SubredditRules:
+    """Provide a set of functions to a Subreddit's rules."""
+
+    @cachedproperty
+    def mod(self) -> _SubredditRulesModeration:
+        """Contain methods to moderate subreddit rules as a whole.
+
+        To add rule ``"No spam"`` to the subreddit ``"NAME"`` try:
+
+        .. code-block:: python
+
+           reddit.subreddit('NAME').rules.mod.add(
+               "No spam",
+               "all",
+               description="Do not spam. Spam bad")
+
+        To move the fourth rule to the first position, and then to move the
+        prior first rule to where the third rule originally was in the
+        subreddit ``"NAME"``:
+
+        .. code-block:: python
+
+            subreddit = reddit.subreddit("NAME")
+            rules = list(subreddit.rules)
+            new_rules = rules[3:4] + rules[1:3] + rules[0:1] + rules[4:]
+            # Alternate: [rules[3]] + rules[1:3] + [rules[0]] + rules[4:]
+            new_rule_list = subreddit.rules.mod.reorder(new_rules)
+
+        """
+        return SubredditRulesModeration(self)
+
+    def __call__(self) -> Iterator:
+        r"""Return a list of :class:`.Rule`\ s (Deprecated)."""
+        warn(
+            "Calling SubredditRules to get a list of rules is deprecated. "
+            "Remove the parentheses to use the iterator.",
+            category=DeprecationWarning,
+            stacklevel=2,
+        )
+        return self.__iter__()
+
+    def __getitem__(self, short_name: Union[str, int, slice]) -> Rule:
+        """Return the Rule for the subreddit with short_name ``short_name``.
+
+        :param short_name: The short_name of the rule, or the rule number.
+
+        .. note:: Rules fetched using a specific rule name are lazy loaded, so
+            you might have to access an attribute to get all of the expected
+            attributes.
+
+        This method is to be used to fetch a specific rule, like so:
+
+        .. code-block:: python
+
+            rule_name = 'No spam'
+            rule = reddit.subreddit('NAME').rules[rule_name]
+            print(rule)
+
+        You can also fetch a numbered rule of a subreddit.
+
+        Rule numbers start at ``0``, so the first rule is at index ``0``, and the
+        second rule is at index ``1``, and so on.
+
+        If a rule of a specific number does not exist, an
+        :py:class:`IndexError` will be thrown.
+
+        .. note:: You can use negative indexes, such as ``-1``, to get the last
+            rule. You can also use slices, to get a subset of rules, such as
+            the last three rules with ``rules[-3:]``.
```

Comment:
I did not expect this to work with just `__getitem__` as implemented, but it does. What is this magic?

Reply:
The call is passed directly to the list index. The slices that are given get passed directly into the list. In other words, if the value is not a string, it's directly passed to the list object.